# Hybrid *lyrata* paper code and data
## By Franklin Brown

**All code supplied by Levi Yant except AFS.R**
### A.
### Code and settings to filter VCF:
#### - Minimum gap size between sites (and reference contig/chromosome ends) = 1000
#### - Lowest acceptable alternate allele frequency per site = 0.02
#### - Highest acceptable ammount of missing data per site = 0.2
python fastStructure_vcf_generation_diversity.py <br>
-i input.vcf <br>
-o output_10kbthin_20PMD_min2pctMAF.vcf <br>
-r C_excelsa_V5.fa.masked.fa.fai <br>
-g 10000 <br>
-f 0.02 <br>
-m 0.2<br>

### B. 
The cohort was generated by selecting columns (individuals) based on presence of matching strings in R. PCA was performed using all individuals from the selected cohort using “glPcaFast” from “Easy_adegenet_script_and_modified_functions.R”, which was then manually filtered to remove the corrupt individuals based on position in the PCA relative to the expected position, producing the final selected cohort of 29 samples for all downstream analyses (Table 1).

### C. 
Neighbour joining trees from distance based analyses were created using “adegenet_Ana_version.R” lines 403 – 517.

### D. 
Allele frequency spectra were created using “AFS.R”.

## Data used in paper
### Reference *lyrata* genome:
NW_003302555.1 Arabidopsis lyrata subsp. lyrata unplaced genomic scaffold ARALYscaffold_1, whole genome shotgun sequence
https://www.ncbi.nlm.nih.gov/refseq/annotation_euk/Arabidopsis_lyrata_subsp._lyrata/101/ <br>

### For original cohort of populations in sections A,B, and C:
- Chrom_1_noSnakemake.lyrata.bipassed.dp.m.bt.1pct.ld_pruned.vcf.gz
- VCF containing 272 individuals

### For creating artificial hybrid in section D:
- lyrata_272_with_some_hybrids.txt.gz <br>
- arenosa_632.txt.gz
